### PR TITLE
use es6 class definitions for all components and upgrade to react 16

### PR DESCRIPTION
**Pull request by Flui**
## Description

- I upgraded react, and react-dom to v16 as react-addons-test-utils do not support react 16 I moved to react-test-renderer.
- I did a refactor on all component to write them in ecma2015 class notation.

## Motivation and Context
The Idea was to keep the project up to date. And as some components were already written in es6, this was also about consistency between the components and to write them in class notation is more convenient as well.

## How Has This Been Tested?
I ran all the tests that has been written for now.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,8 +8,9 @@
     "react-dom": ">=0.14.0 || ^0.14.0-beta3"
   },
   "dependencies": {
-    "create-react-class": "^15.5.2",
-    "prop-types": "^15.5.8"
+    "prop-types": "^15.6.0",
+    "react": "^16.2.0",
```
Comment (wcjordan) [inline package.json]: This shouldn't be a dependency.  It would remove our ability to support old versions.
React-tooltip should also not be a dependency.  Please revert these.

Why did you need to change the prop-types support?  If there's no good reason we should revert that as well.

File: src/FixedDataTable.js [hunk 2]
```diff
@@ -410,21 +406,25 @@ var FixedDataTable = createReactClass({
...
+    showScrollbarY: true,
+    touchScrollEnabled: false,
+    keyboardScrollEnabled: false,
+    keyboardPageEnabled: false,
+    stopScrollPropagation: false
+  };
+
+  constructor(props) {
+    super(props);
+    this._areColumnSettingsIdentical = this._areColumnSettingsIdentical.bind(this);
```
Comment (wcjordan) [inline src/FixedDataTable.js]: You use this syntax and the blah = (args) => { body } arrow syntax.  Seems like we should choose and stick to one.

File: src/FixedDataTableCell.js [hunk 3]
```diff
@@ -90,15 +82,18 @@ var FixedDataTableCell = createReactClass({
...
-      displacement: 0,
-      reorderingDisplacement: 0
-    };
-  },
+  static defaultProps = {
+    align: 'left',
+    highlighted: false,
+  };
+
+  state = {
```
Comment (wcjordan) [inline src/FixedDataTableCell.js]: We should choose this syntax or setting state on this in the constructor and stick to one.

File: src/FixedDataTableColumnReorderHandle.js [hunk 4]
```diff
@@ -16,19 +16,15 @@
 import DOMMouseMoveTracker from 'DOMMouseMoveTracker';
 import Locale from 'Locale';
 import React from 'React';
-import createReactClass from 'create-react-class';
 import PropTypes from 'prop-types';
 import ReactComponentWithPureRenderMixin from 'ReactComponentWithPureRenderMixin';
```
Comment (wcjordan) [inline src/FixedDataTableColumnReorderHandle.js]: can this be removed?

File: src/FixedDataTableColumnResizeHandle.js [hunk 5]
```diff
@@ -17,18 +17,14 @@
 import DOMMouseMoveTracker from 'DOMMouseMoveTracker';
 import Locale from 'Locale';
 import React from 'React';
-import createReactClass from 'create-react-class';
 import PropTypes from 'prop-types';
 import ReactComponentWithPureRenderMixin from 'ReactComponentWithPureRenderMixin';
```
Comment (wcjordan) [inline src/FixedDataTableColumnResizeHandle.js]: Can this be removed?

File: src/Scrollbar.js [hunk 6]
```diff
@@ -13,7 +13,6 @@
 import DOMMouseMoveTracker from 'DOMMouseMoveTracker';
 import Keys from 'Keys';
 import React from 'React';
-import createReactClass from 'create-react-class';
 import PropTypes from 'prop-types';
 import ReactDOM from 'ReactDOM';
 import ReactComponentWithPureRenderMixin from 'ReactComponentWithPureRenderMixin';
```
Comment (wcjordan) [inline src/Scrollbar.js]: Can this be removed?

File: package.json (hunk 1 above)
Comment (Flui) [reply package.json]: ok then I will move this back to devDependencies and revert prop-types version

File: src/FixedDataTable.js (hunk 2 above)
Comment (Flui) [reply src/FixedDataTable.js]: the problem is that react-docgen can't handle static type annotated arrow functions. That throws an parser error. From my point of view using arrow functions is more elegant as it does not add the binding overhead. But as this seems not to work with static type annotations maybe we should use this syntax in general.

File: src/FixedDataTableCell.js (hunk 3 above)
Comment (Flui) [reply src/FixedDataTableCell.js]: The thing is that if the initial state does not depend on the initial props there is no need for a constructor

File: src/FixedDataTableColumnReorderHandle.js (hunk 4 above)
Comment (Flui) [reply src/FixedDataTableColumnReorderHandle.js]: of course

File: src/FixedDataTable.js (hunk 2 above)
Comment (wcjordan) [reply src/FixedDataTable.js]: Ah okay.  Could you add a "// NOTE" comment with that explanation.  Thanks for the explanation.

File: src/FixedDataTableCell.js (hunk 3 above)
Comment (wcjordan) [reply src/FixedDataTableCell.js]: Ah makes sense, thanks for the explanation

**Comment by Flui** [comment]
> I also need to figure out how we can prevent this from causing master and v1.0 from diverging. Do you have details on how you applied this or thoughts on how we can reapply to that branch since a merge to that branch will be difficult?

I am not quite sure what you mean but couldn't we just rebase v1.0 to master. Or what is the issue?

**Comment by wcjordan** [comment]
v1.0 is sizeable rewrite of the Facebook code base to be maintainable and support dynamic row sizes.  I suspect a rebase will include a significant number of merge conflicts.  You're welcome to give it a shot and see how it goes.

**Comment by Flui** [comment]
I think the way to go is to rebase master to v1.0.0-beta.4 and afterwards I would rebase refactor/react-components to master

**Comment by Flui** [comment]
Then I will try to apply my changes to v1 and change the target of the pull request to v1. Yeah I also integrated jest to add snapshottesting https://github.com/schrodinger/fixed-data-table-2/pull/297

**Comment by Flui** [comment]
Ok I did a rebase to v1 and I also removed the commit where I upgraded to react 16 as this can be done in another pull request.

**Comment by wcjordan** [comment]
looks good to me, thanks for the changes!

Travis seems to not work on Node 4, which I'm fine with dropping.  Could you update [.travis.yml](https://github.com/schrodinger/fixed-data-table-2/blob/master/.travis.yml) to include 6,7,8, & 9 instead of 4,5, & 6?
